Update Native AOT documentation (#128021)
dotnet/runtime: docs/design/coreclr/botr/managed-type-system.md

Signature variables represent variables that can be substituted by other types within the system. They differ from generic parameters (because e.g. they don't have constraints or variance). They are simply placeholders to be replaced by other types as part of a process called instantiation. Signature variables have an index that refers to a position within the instantiation context.

Consider a class `Foo<T>` that has a method `Bar<U>(T x, U y)`. When IL references this method in a signature, `T` becomes `!0` (a SignatureTypeVariable meaning "the first type argument of the declaring type") and `U` becomes `!!0` (a SignatureMethodVariable meaning "the first type argument of the declaring method").

## Other type system classes

Each use of a type system starts with creating a type system context. A type system context represents a type universe across which all types share reference identity (two `TypeDesc` objects represent identical types if and only if they are the same object instance). Type system context is used to resolve all modules and constructed types within the universe. It's not legal to create new instances of constructed types outside of the type system context.

> The type system contexts all share the same base (`TypeSystemContext`), but each configures different pluggable algorithms, loads assemblies differently, and may support different synthetic types. For example, the NativeAOT compiler context and the ReadyToRun (crossgen2) compiler context use different field layout algorithms to compensate for small differences such as whether the `MethodTable` pointer in `System.Object` is a regular pointer field (native AOT) or just "vacated pointer-sized space" (CoreCLR VM type system).

Reviewer: ```suggestion
The type system contexts all share the same base (`TypeSystemContext`), but each configures different pluggable algorithms, loads assemblies differently, and may support different synthetic types. For example, the NativeAOT compiler context and the ReadyToRun (crossgen2) compiler context use different field layout algorithms to compensate for small differences such as whether the `MethodTable` pointer in `System.Object` is a regular pointer field (native AOT) or just "vacated pointer-sized space" (CoreCLR type system).
```
Use consistent names throughout the doc

Signature variables represent variables that can be substituted by other types within the system. They differ from generic parameters (because e.g. they don't have constraints or variance). They are simply placeholders to be replaced by other types as part of a process called instantiation. Signature variables have an index that refers to a position within the instantiation context.

> Consider a class `Foo<T>` that has a method `Bar<U>(T x, U y)`. When IL references this method in a signature, `T` becomes `!0` (a SignatureTypeVariable meaning "the first type argument of the enclosing type") and `U` becomes `!!0` (a SignatureMethodVariable meaning "the first type argument of the method").

Reviewer 2: ```suggestion
Consider a class `Foo<T>` that has a method `Bar<U>(T x, U y)`. When IL references this method in a signature, `T` becomes `!0` (a SignatureTypeVariable meaning "the first type argument of the declaring type") and `U` becomes `!!0` (a SignatureMethodVariable meaning "the first type argument of the declaring method").
```

## Other type system classes

Each use of a type system starts with creating a type system context. A type system context represents a type universe across which all types share reference identity (two `TypeDesc` objects represent identical types if and only if they are the same object instance). Type system context is used to resolve all modules and constructed types within the universe. It's not legal to create new instances of constructed types outside of the type system context.

> The type system contexts all share the same base (`TypeSystemContext`), but each configures different pluggable algorithms, loads assemblies differently, and may support different synthetic types. For example, the NativeAOT compiler context and the ReadyToRun (crossgen2) compiler context use different field layout algorithms — NativeAOT has full control over type layouts since it produces the final binary, while ReadyToRun must match the layout decisions that the CoreCLR runtime will make at load time.

Reviewer 2: ```suggestion
The type system contexts all share the same base (`TypeSystemContext`), but each configures different pluggable algorithms, loads assemblies differently, and may support different synthetic types. For example, the NativeAOT compiler context and the ReadyToRun (crossgen2) compiler context use different field layout algorithms to compensate for small differences such as whether the `MethodTable` pointer in `System.Object` is a regular pointer field (native AOT) or just "vacated pointer-sized space" (CoreCLR VM type system).
```

Similar to `TypeDesc` hierarchy mentioned some sections above, `MethodDesc` follows the same extensible hierarchy pattern and represents all methods in the type system. Some of its subclasses include `EcmaMethod` (methods that read from ECMA-335 metadata), `ArrayMethod` (synthesized array methods), `InstantiatedMethod` (generic method instantiations like `Foo<int>`), and `ILStubMethod` (compiler-generated stubs for scenarios like P/Invoke marshalling).

A `ModuleDesc` describes a single module which can optionally implement `IAssemblyDesc` interface if the module is an assembly. `ModuleDesc` is typically the owner of the type/method/field definitions within the module. It's the responsibility of the `ModuleDesc` to maintain the reference identity of those.

## Accessing ECMA-335 metadata

> While the type system classes backed by ECMA-335 metadata (`EcmaMethod`, `EcmaType`, `EcmaField`) expose commonly needed information through their properties, compiler code frequently needs to access metadata that the type system doesn't project — custom attributes, parameter details, generic constraints, and so on. To achieve that, there is the following pattern:

Reviewer 2: ```suggestion
While the type system classes backed by ECMA-335 metadata (`EcmaMethod`, `EcmaType`, `EcmaField`) expose commonly needed information through their properties, compiler code sometimes needs to access metadata that the type system doesn't project - custom attributes, parameter details, generic constraints, and so on. To achieve that, there is the following pattern:
```

MethodDefinition methodDef = reader.GetMethodDefinition(methodHandle);
```

Each step in this pattern serves a distinct purpose.

`MetadataReader` is the cursor over the metadata tables of a single module/assembly. It owns the underlying memory block containing the metadata blob and the parsed table structures. Every `EcmaModule` carries one `MetadataReader`, and the `Ecma*` type system classes expose it through a property.

A handle (such as `MethodDefinitionHandle`, `TypeDefinitionHandle`, or `FieldDefinitionHandle`) is just a 4-byte value that identifies a specific row but carries no data on its own. To get information you must pass it to a `MetadataReader`. The `Ecma*` type system classes expose their handle through the `Handle` property.

A definition struct (such as `MethodDefinition`, `TypeDefinition`, or `FieldDefinition`) is a decoded view of a single metadata table row. After passing the handle to the reader, the resulting struct provides access to the row's columns. Its metadata accessors (attributes, name, signature, etc.) return either a primitive, a flags enum, or another handle that must be decoded through the same reader — no allocations involved.

> This lazy, zero-allocation pattern is what allows the `Ecma*` types to remain thin wrappers that cache only frequently accessed data (such as the parsed method signature and name) while decoding everything else on demand through the reader.

Reviewer 2: I would skip this part, this is basically documenting the System.Reflection.Metadata APIs.
